mapillary.py

```diff
@@ -0,0 +1,138 @@
+
+    finished_bbox_sections = multiprocessing.Value('i', 0)
+    with multiprocessing.Pool(initializer=initialize_multiprocess,
+                              initargs=(output_dir, output_tmp_dir, traces_source, finished_bbox_sections),
+                              processes=processes) as pool:
+        result = pool.map_async(pull_and_save_trace_for_bbox, bbox_sections)
+
+        print('Placing {} results in {}...'.format(len(bbox_sections), output_dir))
+        progress = 0
+        increment = 5
+        while not result.ready():
+            result.wait(timeout=5)
+            next_progress = int(finished_bbox_sections.value / len(bbox_sections) * 100)
+            if int(next_progress / increment) > progress:
+                print('Current progress: {}%'.format(next_progress))
+                progress = int(next_progress / increment)
+        if progress != 100 / increment:
+            print('Current progress: 100%')
+
+        # TODO: Delete the tmp dir after run?
+
+
+def to_bbox(llo, lla, mlo, mla):
+    return ','.join([str(llo), str(lla), str(mlo), str(mla)])
+
+
+def initialize_multiprocess(global_output_dir_: str, global_output_tmp_dir_: str, global_traces_source_: any,
+                            finished_bbox_sections_: multiprocessing.Value):
+    # For persistent connections
+    global session
+    session = requests.Session()
+
+    # So each process knows the output / tmp dirs
+    global global_output_dir
+    global_output_dir = global_output_dir_
+    global global_output_tmp_dir
+    global_output_tmp_dir = global_output_tmp_dir_
+
+    # So each process knows the conf provided
+    global global_traces_source
+    global_traces_source = global_traces_source_
+
+    global finished_bbox_sections
+    finished_bbox_sections = finished_bbox_sections_
+
+
+def pull_and_save_trace_for_bbox(bbox: str) -> None:
+    try:
+        # The file on disk where we will store trace data
+        result_filename = os.path.join(global_output_dir, bbox + '.pickle')
+
+        if os.path.exists(result_filename):
+            print('Seq for bbox={} already exists on disk! Skipping...'.format(bbox))
+            with finished_bbox_sections.get_lock():
+                finished_bbox_sections.value += 1
+            return
+
+        # We haven't pulled API trace data for this bbox section yet
+        trace_data = trace_data_request(session, bbox, global_traces_source)
+
+        # Avoids potential partial write issues by writing to a temp file and then as a final operation, then renaming
+        # to the real location
+        temp_filename = os.path.join(global_output_tmp_dir, bbox + '.pickle')
+        pickle.dump(trace_data, open(temp_filename, 'wb'))
+        os.rename(temp_filename, result_filename)
+
+        with finished_bbox_sections.get_lock():
+            finished_bbox_sections.value += 1
+    except Exception as e:
+        print('ERROR: Failed to pull trace data: {}'.format(repr(e)))
```

Comment:
@kevinkreiser This is the method that is called in async parallel. Given a `bbox_section` to work on, it checks if it exists and if not calls the API as usual.